tests/test_cache.py

```diff
@@ -1,504 +1,372 @@
+    assert mock_session.get(MOCKED_URL, params={"a": "b"}).from_cache is False
+    assert mock_session.get(MOCKED_URL, params=dict(sorted(params.items()))).from_cache is True
+
+    class UserSubclass(dict):
+        def items(self):
+            return sorted(super(UserSubclass, self).items(), reverse=True)
+
+    params["z"] = "5"
+    custom_dict = UserSubclass(params)
+    assert mock_session.get(MOCKED_URL, params=custom_dict).from_cache is False
+    assert mock_session.get(MOCKED_URL, params=custom_dict).from_cache is True
+
+
+def test_normalize_post_data(mock_session):
+    params = {"a": "a", "b": ["1", "2", "3"], "c": "4"}
+    assert mock_session.post(MOCKED_URL, data=params).from_cache is False
+    assert mock_session.post(MOCKED_URL, data=params).from_cache is True
+    assert mock_session.post(MOCKED_URL, data=sorted(params.items())).from_cache is True
+    assert mock_session.post(MOCKED_URL, data=sorted(params.items(), reverse=True)).from_cache is False
+
+
+def test_delete_response(mock_session):
+    mock_session.get(MOCKED_URL)
+    mock_session.cache.delete_url(MOCKED_URL)
+    assert not mock_session.cache.has_url(MOCKED_URL)
+    mock_session.cache.delete_url(MOCKED_URL)  # Should fail silently
+
+
+# TODO: Better mocking for redirects
+def test_delete_redirect(mock_session):
+    response_key = mock_session.cache._url_to_key(MOCKED_URL)
+    redirect_key = mock_session.cache._url_to_key(MOCKED_URL_REDIRECT)
+    mock_session.get(MOCKED_URL)
+    mock_session.cache.keys_map[redirect_key] = response_key
+
+    mock_session.cache.delete_url(MOCKED_URL_REDIRECT)
+    assert mock_session.cache.has_url(MOCKED_URL)
+    assert not mock_session.cache.has_url(MOCKED_URL_REDIRECT)
+
+
+# TODO
+def test_delete_history(mock_session):
+    pass
+
+
+def test_response_defaults(mock_session):
+    """Both cached and new responses should always have the following attributes"""
+    mock_session.expire_after = datetime.utcnow() + timedelta(days=1)
+    response_1 = mock_session.get(MOCKED_URL)
+    response_2 = mock_session.get(MOCKED_URL)
+    response_3 = mock_session.get(MOCKED_URL)
+
+    assert response_1.created_at is None
+    assert response_1.expires is None
+    assert response_1.from_cache is False
+    assert response_1.is_expired is False
+
+    assert isinstance(response_2.created_at, datetime)
+    assert isinstance(response_2.expires, datetime)
+    assert response_2.created_at == response_3.created_at
+    assert response_2.expires == response_3.expires
+    assert response_2.from_cache is response_3.from_cache is True
+    assert response_2.is_expired is response_3.is_expired is False
+
+
+def test_include_get_headers(mock_session):
+    """With include_get_headers, requests with different headers should have different cache keys"""
+    mock_session.cache._include_get_headers = True
+    headers_list = [{'Accept': 'text/json'}, {'Accept': 'text/xml'}, {'Accept': 'custom'}, None]
```

Comment:
Are headers normalized as well? In other words, what about:
```
headers = {'Accept': 'text/json', 'Custom': 'abc'}
headers = {'Custom': 'abc', 'Accept': 'text/json'}
```

Reply:
Good point, I added a test for that. And I'll look into that a little more thoroughly for issue #80.